_posts/2017-11-30-totw-119.md

```diff
@@ -0,0 +1,264 @@
+---
+title: "Tip of the Week #119: Using-declarations and namespace aliases"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/119
+type: markdown
+order: "119"
+---
+
+Originally posted as totw/119 on ORIGINAL DATE
+
+By Thomas K&ouml;ppe (tkoeppe@google.com)
+
+This tip gives a simple, robust recipe for writing *using-declarations* and
+namespace aliases in `.cc` files that avoids subtle pitfalls.
+
+Before we dive into the details, here is an example of the recipe in action:
+
+```c++
+namespace example {
+namespace makers {
+namespace {
+
+using ::otherlib::BazBuilder;
+using ::mylib::BarFactory;
+namespace abc = ::applied::bitfiddling::concepts;
+
+// Private helper code here.
+
+}  // namespace
+
+// Interface implementation code here.
+
+}  // namespace makers
+}  // namespace example
+```
+
+Remember that everything in this tip applies only to `.cc` files, since we
+[never put convenience aliases in header
+files](http://google.github.io/styleguide/cppguide.html#Aliases). Such
+aliases are a convenience for the implementer (and the reader of an
+implementation), not an exported facility. (Names that *are* part of the
+exported API may of course be declared in headers.)
+
+## Summary
+
+*   Never declare namespace aliases or convenience <em>using-declaration</em>s
+    at namespace scope in header files, only in `.cc` files.
+
+*   Declare namespace aliases and <em>using-declaration</em>s inside the
+    innermost namespace, whether named or anonymous. (Do not add an anonymous
+    namespace just for this purpose.)
+
+*   When declaring namespace aliases and <em>using-declaration</em>s, use fully
+    qualified names (with leading `::`) unless you are referring to a name
+    inside the current namespace.
+
+*   For other uses of names, avoid fully qualifying when reasonable; see
+    [TotW 130](/tips/130).
+
+(Remember that you can always have a local namespace alias or
+*using-declaration* in a block scope, which can be handy in header-only
+libraries.)
+
+## Background
+
+C++ organizes names into *namespaces*. This crucial facility allows code bases
+to scale by keeping ownership of names local and not requiring non-local
```

Comment:
"not requiring non-local" reads weird. Though "not requiring global" is incorrect.

Reword as:
"by keeping ownership of names local and avoiding name collisions in other scopes." (?)